src/hooks/auditoriumSections.ts

```diff
@@ -0,0 +1,134 @@
+import { User } from "types/User";
+
+import { setGridData } from "api/profile";
+
+import {
+  currentAuditoriumSectionsSelector,
+  currentAuditoriumSectionsByIdSelector,
+} from "utils/selectors";
+import { WithId } from "utils/id";
+import { getPositionHash } from "utils/auditorium";
+
+import { useSelector } from "./useSelector";
+import { useFirestoreConnect, isLoaded } from "./useFirestoreConnect";
+import { useRecentVenueUsers } from "./users";
+import { useUser } from "./useUser";
+import { useCallback, useMemo } from "react";
+
+const useConnectAuditoriumSections = (venueId?: string) => {
+  useFirestoreConnect(() => {
+    if (!venueId) return [];
+
+    return [
+      {
+        collection: "venues",
+        doc: venueId,
+        subcollections: [{ collection: "sections" }],
+        storeAs: "currentAuditoriumSections",
+      },
+    ];
+  });
+};
+
+export interface UseAuditoriumSectionProps {
+  sectionId: string;
+  venueId?: string;
+}
+
+export const useAuditoriumSection = ({
+  venueId,
+  sectionId,
+}: UseAuditoriumSectionProps) => {
+  useConnectAuditoriumSections(venueId);
+
+  const { userWithId } = useUser();
+  const userId = userWithId?.id;
+
+  const sections = useSelector(currentAuditoriumSectionsByIdSelector);
+  const section = sections?.[sectionId];
+
+  const seatedUsers = useSectionSeatedUsers(venueId, sectionId);
+
+  const seatedUsersByHash = useMemo(
+    () =>
+      seatedUsers.reduce<Record<string, WithId<User> | undefined>>(
+        (acc, user) => {
+          if (!venueId) return acc;
+
+          const takenRow = user.data?.[venueId].row;
+          const takenColumn = user.data?.[venueId].column;
+
+          if (takenRow === undefined || takenColumn === undefined) return acc;
+
+          const positionHash = getPositionHash({
+            row: takenRow,
+            column: takenColumn,
+          });
+
+          return { ...acc, [positionHash]: user };
+        },
+        {}
+      ),
+    [seatedUsers, venueId]
+  );
+
+  const getUserBySeat = useCallback(
+    ({ row, column }: { row: number; column: number }) =>
+      seatedUsersByHash && seatedUsersByHash[getPositionHash({ row, column })],
+    [seatedUsersByHash]
+  );
+
+  const takeSeat = useCallback(
+    ({ row, column }: { row: number; column: number }) => {
+      if (!sectionId || !venueId || !userId) return;
+
+      setGridData({
+        venueId,
+        userId,
+        seatOptions: { sectionId, row, column },
+      });
```

Comment:
I don't think we should add the error handler in every place we use the function. A better approach would be to handle it in the api/profile, since the shown error should be the same.

Added promise return in https://github.com/sparkletown/sparkle/pull/1251/commits/da3f021502e1e26ea4bbe80350096b14cb2bb6cc

[OOS] Toast messages: https://github.com/sparkletown/internal-sparkle-issues/issues/630

Didn't understand this part `And then we should also properly pass that Promise type 'out' in the types of takeSeat`

Reply:
> I don't think we should add the error handler in every place we use the function. A better approach would be to handle it in the api/profile, since the shown error should be the same.

I guess that depends on how we choose to display the error. IMO displaying it to the user is likely a separate case from the API layer. It's a UI layer concern, and will depend on where the function is called/etc. That's why I said we should handle it 'everywhere it is called' (eg. in the component/etc that uses it)

Reply:
> > And then we should also properly pass that Promise type 'out' in the types of takeSeat
>
> Didn't understand this part

I believe my request here was just what you did with this:

> Added promise return in da3f021

Reply:
> I guess that depends on how we choose to display the error. IMO displaying it to the user is likely a separate case from the API layer. It's a UI layer concern, and will depend on where the function is called/etc. That's why I said we should handle it 'everywhere it is called' (eg. in the component/etc that uses it)

Let's continue this conversation in the issue 630